test_proxy/test_proxy.py

```diff
@@ -0,0 +1,193 @@
+objects cannot be used in the same process, so we had to make use of the
+multiprocessing module to allow them to work together.
+"""
+
+import multiprocessing
+import argparse
+import sys
+import os
+sys.path.append("handlers")
+
+
+def grpc_server_process(request_q, queue_pool, port=50055):
+    """
+    Defines a process that hosts a grpc server
+    proxies requests to a client_handler_process
+    """
+    sys.path.append("protos")
+    from concurrent import futures
+
+    import grpc
+    import test_proxy_pb2_grpc
+    import grpc_handler
+
+    # Start gRPC server
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    test_proxy_pb2_grpc.add_CloudBigtableV2TestProxyServicer_to_server(
+        grpc_handler.TestProxyGrpcServer(request_q, queue_pool), server
+    )
+    server.add_insecure_port("[::]:" + port)
+    server.start()
+    print("grpc_server_process started, listening on " + port)
+    server.wait_for_termination()
+
+
+async def client_handler_process_async(request_q, queue_pool, use_legacy_client=True):
+    """
+    Defines a process that recives Bigtable requests from a grpc_server_process,
+    and runs the request using a client library instance
+    """
+    import base64
+    import re
+    import asyncio
+    import warnings
+    import client_handler_data
+    import client_handler_legacy
+    warnings.filterwarnings("ignore", category=RuntimeWarning, message=".*Bigtable emulator.*")
+
+    def camel_to_snake(str):
+        return re.sub(r"(?<!^)(?=[A-Z])", "_", str).lower()
+
+    def format_dict(input_obj):
+        if isinstance(input_obj, list):
+            return [format_dict(x) for x in input_obj]
+        elif isinstance(input_obj, dict):
+            return {camel_to_snake(k): format_dict(v) for k, v in input_obj.items()}
+        elif isinstance(input_obj, str):
+            # check for time encodings
+            if re.match("^[0-9]+s$", input_obj):
+                return int(input_obj[:-1])
+            # check for encoded bytes
+            if re.match("^[A-Za-z0-9+/=]+$", input_obj):
+                try:
+                    decoded_str = base64.b64decode(input_obj)
+                    # if the string contains non-ascii bytes, raise exception
+                    decoded_str.decode("ascii")
+                    return decoded_str
+                except Exception:
+                    pass
+            # check for int strings
+            try:
+                return int(input_obj)
+            except (ValueError, TypeError):
+                return input_obj
+        else:
+            return input_obj
+
+    # Listen to requests from grpc server process
+    print_msg = "client_handler_process started"
+    if use_legacy_client:
+        print_msg += ", using legacy client"
+    print(print_msg)
+    client_map = {}
+    background_tasks = set()
+    while True:
+        if not request_q.empty():
```

Comment:
I'm still trying to understand the architecture of this PR. Why is the functionality of CreateClient, CloseClient and RemoveClient handled here instead of being delegated down to the next level like all the other functions sort of like how the node.js test proxy works? (ie. https://github.com/googleapis/nodejs-bigtable/blob/79035298bbf663c98d46c95021f7c944ef2c97f2/testproxy/services/index.js#L38-L47)

Reply:
Yes, it is annoyingly complicated because of a quirk of python protobuf. We can't load gapic proto objects and raw proto objects in the same Python instance, so we need to do a complex multi-processing system that passes messages through a queue to accomplish anything here.

The original test_proxy PR is [here](https://github.com/googleapis/python-bigtable/pull/836), which has more discussuion if you're interested, and I tried to describe some of the architecture in [the docstrings](https://github.com/daniel-sanche/python-bigtable/blob/50648705b88b75219ae418979427dabdbe1b1cec/test_proxy/test_proxy.py#L22). But I suggest you ignore all the test_proxy code that was already reviewed there, this is just moving the same code into the `main` branch